### ISSUE TITLE
Expose more structured logging capabilities to `{{ log() }}`

### ISSUE THREAD

**Issue by jtcohen6** [dbt-labs/dbt-core#4184]
Picking up from https://github.com/dbt-labs/dbt-core/pull/4164#discussion_r739298765

By rough estimate, the most critical logging is:
- 70% in dbt-core
- 10% in dbt adapter plugins, i.e. for debugging connections
- 20% calls to the context macro `{{ log() }}`

Why so much in the last category? This could be logging from:
- [materializations](https://github.com/dbt-labs/dbt-bigquery/blob/bb9caab8fd771d72bafdcfffb1ae72459ad704a2/dbt/include/bigquery/macros/materializations/table.sql#L24), built-in or custom
- [other macros](https://github.com/dbt-labs/dbt-utils/blob/main/macros/jinja_helpers/log_info.sql) that wrap `log()`
- [complex operations](https://github.com/dbt-labs/dbt-external-tables/blob/master/macros/common/stage_external_sources.sql#L64), e.g. those defined in packages
- any custom user-space code

Based on the implementation of structured logging that will be included in v1, these log messages will be passed along as relatively unstructured events, `MacroEventInfo` or `MacroEventDebug`, with just one string property, `msg`.

As structured logs become an increasingly important interface for interacting with / listening to dbt, we should seek to open up more capabilities to end users who want to hook into more powerful logging from the macro/materialization context. I could see us doing that by:
- Adding a new argument to the `log()` macro that enables users to pass along any key-value pairs they want, and a templated message (Jinja expression or f-string) using those keys/values
- Adding complementary information gleaned from the Jinja context—if it's even available!—e.g. the parent macro/frame calling this `log()` message

**Comment by nathaniel-may** [comment]
I love the idea of exposing some of this new functionality via our jinja interface. I do want to ask if some of the common use cases for `{{ log() }}` like materializations should just be implemented in core so they don't have to be written by users. For the remaining calls to `{{ log() }}` I like the direction you're heading by allowing users to construct messages at runtime with functions. I suspect this is possible today, but might look less graceful than it could. Is there any log functionality that is simply not possible today?

**Comment by jtcohen6** [comment]
The context available to users is the same as that available to dbt-core's built-in materializations today. It's a good callout, we could think about supporting other types of logging/event functions in the Jinja context, or optional arguments to `{{ log() }}` that would lead it to call (e.g.) a `MaterializationEventInfo` instead of `MacroEventInfo`.

From my vantage, I was thinking that the simplest way would be to add an argument to `{{ log() }}` along the lines of:
```python
@dataclass
class MacroEventInfo(InfoLevel, CliEventABC):
    msg: str
    properties: dict

    def cli_msg(self) -> str:
        return self.msg
```
```sql
dict = {thing_a: x, thing_b: y}
message = "Could not do {{ thing_a }} because {{ thing_b }}"
{{ log(message, info = true, properties = dict) }}
```
To achieve the equivalent of:
```python
class ThisSpecificCustomEvent(...):
  thing_a: str
  thing_b: str

def cli_msg(self) -> str:
  return(
    f"Could not {self.thing_a} because {self.thing_b}"
  )
```
And be able to use the additional structure of `properties` accordingly

**Comment by nathaniel-may** [comment]
Yeah I can see how that would work, and it seems consistent to use Jinja here as opposed to python f-string syntax for instance.

**Comment by jtcohen6** [comment]
Another version of this that we discussed, following the same pseudo-code above:
```sql
{% set thing_a = ... %}
{% set thing_b = ... %}
{{ fire_event(event_type='MyCoolEvent', thing_a=thing_a, thing_b=thing_b }}
```
Wherein `fire_event()` is a Jinja macro / context member that, in turn, calls `fire_event(EventType(**kwargs))`.

**Comment by swanderz** [comment]
as someone who spends way too much time debugging jinja, I'd like to "yes, and..." this idea by saying it'd be great to have to log message return:
- a timestamp of the same format as the rest of the logs,
- the callstack of jinja macros from where it's called (a la [Delgan/loguru](https://github.com/Delgan/loguru)), and
- a shorthand way to print the name of var in addition to it's value? akin to the fstring impl mentioned in #3911

**Comment by nathaniel-may** [comment]
Closing in favor of #5325. Since this is going to be a part of that effort.
